storm/src/main/storm/mesos/schedulers/OfferResources.java

```diff
@@ -0,0 +1,181 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * <p/>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p/>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package storm.mesos.schedulers;
+
+import com.google.common.base.Joiner;
+import org.apache.log4j.Logger;
+import org.apache.mesos.Protos;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class OfferResources {
+  private final Logger log = Logger.getLogger(OfferResources.class);
+
+  public enum ResourceType {
+    CPU,
+    MEM,
+    PORT
+  }
+
+  private class PortRange {
+    public long begin;
+    public long end;
+
+    public PortRange(long begin, long end) {
+      this.begin = begin;
+      this.end = end;
+    }
+  }
+
+  Protos.Offer offer;
+  Protos.OfferID offerId;
+  String hostName;
+  double mem;
+  double cpu;
+
+  List<PortRange> portRanges = new ArrayList<>();
+
+  public void addPortRanges(Protos.Value.Ranges ranges) {
+    for (Protos.Value.Range r : ranges.getRangeList()) {
+      this.portRanges.add(new PortRange(r.getBegin(), r.getEnd()));
+    }
+  }
+
+  public OfferResources(Protos.Offer offer) {
+    this.offer = offer;
+    this.offerId = offer.getId();
+    double offerMem = 0;
+    double offerCpu = 0;
+    Protos.Value.Ranges portRanges = null;
+
+    String hostName = offer.getHostname();
+    for (Protos.Resource r : offer.getResourcesList()) {
+      if (r.getName().equals("cpus")) {
+        offerCpu = r.getScalar().getValue();
+      } else if (r.getName().equals("mem")) {
+        offerMem = r.getScalar().getValue();
+      } else if (r.getName().equals("ports")) {
+        portRanges = r.getRanges();
+      }
+    }
+
+    this.hostName = hostName;
+    this.mem = offerMem;
+    this.cpu = offerCpu;
+    if ((portRanges != null) && (!portRanges.getRangeList().isEmpty())) {
+      this.addPortRanges(portRanges);
+    }
+  }
+
```

Comment:
Looks like you stripped out any awareness of roles here.  If a cluster has a role with reserved resources for storm, an `offer.getResourceList` will have two `Resource` of type `SCALAR` with name `cpus`, `mem`, `ports`, etc, one from the default role and one with the named role.  Here your `for` loop is simply overriding the value, so you'll get the last `Resource` of the name.

It will also complicate the TaskInfo stage a bit, as you'll need to add two resources if the total value the task needs spans the Resources of the offer.

 @brndnmtthws just added resource role awareness into Storm, I added this to Myriad a while ago as well if you have questions.